torch/lib/c10d/test/ProcessGroupMPITest.cpp

```diff
@@ -0,0 +1,98 @@
+#include <cstdlib>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <thread>
+
+#include "ProcessGroupMPI.hpp"
+
+void testAllreduce(int iter = 1000) {
+  auto pg = c10d::ProcessGroupMPI::createProcessGroupMPI();
+  // Generate inputs
+  std::vector<std::vector<at::Tensor>> allTensors(iter);
+  for (auto i = 0; i < iter; ++i) {
+    auto tensor = at::ones(at::CPU(at::kFloat), {16, 16}) * i;
+    allTensors[i] = std::vector<at::Tensor>({tensor});
+  }
+
+  std::vector<std::shared_ptr<::c10d::ProcessGroup::Work>> works;
+  for (auto& tensors : allTensors) {
+    // Kick off work
+    std::shared_ptr<::c10d::ProcessGroup::Work> work = pg->allreduce(tensors);
+    works.push_back(std::move(work));
+  }
+
+  for (auto& work : works) {
+    // Wait for work to complete
+    if (!work->wait()) {
+      std::cerr << "Exception received: " << work->exception().what()
+                << std::endl;
+      pg->abort();
+    }
+  }
+
+  // Get the world size
+  auto worldSize = pg->getSize();
+
+  // Verify outputs
+  for (size_t i = 0; i < iter; ++i) {
+    const auto expected = worldSize * i;
+    auto data = allTensors[i][0].data<float>();
+    for (auto i = 0; i < allTensors[i][0].numel(); ++i) {
+      if (data[i] != expected) {
+        throw std::runtime_error("BOOM!");
+      }
+    }
+  }
+}
+
+void testBroadcast(int iter = 10000) {
+  auto pg = c10d::ProcessGroupMPI::createProcessGroupMPI();
+  // Generate inputs
+  std::vector<std::vector<at::Tensor>> allTensors(iter);
+
+  for (auto i = 0; i < iter; ++i) {
+    if (pg->getRank() == 0) {
+      auto tensor = at::ones(at::CPU(at::kFloat), {16, 16}) * i;
+      allTensors[i] = std::vector<at::Tensor>({tensor});
+    } else {
+      auto tensor = at::zeros(at::CPU(at::kFloat), {16, 16});
+      allTensors[i] = std::vector<at::Tensor>({tensor});
+    }
+  }
+
+  std::vector<std::shared_ptr<::c10d::ProcessGroup::Work>> works;
+  for (auto& tensors : allTensors) {
+    // Kick off work
+    std::shared_ptr<::c10d::ProcessGroup::Work> work = pg->broadcast(tensors);
+    works.push_back(std::move(work));
+  }
+
+  for (auto& work : works) {
+    // Wait for work to complete
+    if (!work->wait()) {
+      std::cerr << "Exception received: " << work->exception().what()
+                << std::endl;
+      pg->abort();
+    }
+  }
+
+  // Verify outputs
+  for (size_t i = 0; i < iter; ++i) {
+    const auto expected = i;
+    auto data = allTensors[i][0].data<float>();
+    for (auto i = 0; i < allTensors[i][0].numel(); ++i) {
+      if (data[i] != expected) {
+        throw std::runtime_error("BOOM!");
+      }
+    }
+  }
+}
+
+int main(int argc, char** argv) {
+  testAllreduce();
+  testBroadcast();
```

Comment:
You list mpirun needed to run this test. What happens if you run make test in the build dir?

I think the test should be runnable as part of a larger suite. Otherwise this test will never be run. This can be done for example by checking if MPI specific env vars are set, and if not, assume it is running by itself, and exec into mpirun.

Reply:
Can we do this in the CI test script for this.  Would be much easier to launch by python? :)

Reply:
Ok, now this is added